### ISSUE TITLE
Question re: iterations when tuning

### ISSUE THREAD

**Issue by anpatton** [stanfordmlgroup/ngboost#105]
Hello,

If this is too questiony as opposed an issue, feel free to delete, but this seems to be the best place to ask. I was running the toy examples on the explanation page after running into some issues with my model. @kmedved

Below is the verbose=True output from the most basic use case shown.

`[iter 0] loss=3.6260 val_loss=0.0000 scale=0.5000 norm=3.3052
[iter 100] loss=3.0573 val_loss=0.0000 scale=1.0000 norm=3.7136
[iter 200] loss=2.4105 val_loss=0.0000 scale=2.0000 norm=4.0158
[iter 300] loss=2.0426 val_loss=0.0000 scale=1.0000 norm=1.6342
[iter 400] loss=1.8848 val_loss=0.0000 scale=1.0000 norm=1.4886`

However, when I add any sort of tuning, even the one on the vignette, I get the following, which produces predictions with much more error that are in an extremely narrow band.

`[iter 0] loss=3.6708 val_loss=0.0000 scale=0.5000 norm=3.5097
[iter 0] loss=3.6525 val_loss=0.0000 scale=0.5000 norm=3.3920
[iter 0] loss=3.6471 val_loss=0.0000 scale=0.5000 norm=3.3694
[iter 0] loss=3.6346 val_loss=0.0000 scale=0.5000 norm=3.3569
[iter 0] loss=3.6596 val_loss=0.0000 scale=0.5000 norm=3.4142
[iter 0] loss=3.6532 val_loss=0.0000 scale=0.5000 norm=3.4083`

Not sure where I'm making a mistake or just misunderstanding.
[
Link to Colab with example code.](https://colab.research.google.com/drive/1L4SXOFPqox0PkPAoeELXF_rGPJM0Wofj)

**Comment by alejandroschuler** [comment]
Hey @anpatton I think the issue is that you have `'n_estimators': [20, 40]` in the tuning grid, at least in the example in the colab notebook. So the models are relatively underfit, which explains why the predictions are in a narrow bound. When I remove that from the tuning grid I get:
![image](https://user-images.githubusercontent.com/7213095/76153550-3a660180-6082-11ea-8f6e-aae0297eb744.png)

I'm glad you asked this, though, because this points out that it's hard right now to do grid search tuning for tree depth, learning rate, etc. while doing early stopping to tune the number of estimators at the same time. You can tune the number of estimators with a grid search, but it's needlessly computationally intense since you train the 20-tree model, then retrain the 40-tree model from scratch, etc. I'll see if I can come up with a way to combine these two options for tuning in a little bit and update the documentation to demonstrate how to do it properly.

**Comment by kmedved** [comment]
@alejandroschuler with respect to how to tune n_estimators, my approach, stolen from a variety of Kagglers, is to not tune the `n_estimators` parameter and instead set it at some extremely high number (say 10,000), and then rely on early stopping to essentially tune that parameter for me, setting the `n_estimators` in the final fit model at the value of `best_val_loss_itr`.

However, when I try that approach with ngboost, it doesn't seem to work. For some reason, using the Boston data as an example, early stopping pretty routinely kicks in at around ~150-200 estimators, but when I try to actually fit the model and cross validate, I get quite poor results with so few estimators.

I've dug into this a bit, and I've found what seems to be somewhat unusual behavior in ngboost when validation data is passed into `.fit()`, even **without** early stopping. I've put up an example notebook on Colab here so you can see: https://colab.research.google.com/drive/12Cf5WzraW3Z4B2G4QwFNr1Cyh9FV9duO#scrollTo=YKlhAGKogyK5

In `NGBoost Model 1` (Cell 4), I run NGBoost for 1000 estimators using otherwise default parameters. In `NGBoost Model 2` (Cell 5), I run the same code with the same data, except I pass validation data to the model as well (but no early stopping). In other GBDT models, I would expect to see no change here, but actually the result is a pretty different model with a materially **worse** MSE. This result is perplexing, because I'd expect this to be the same model as `NGBoost Model 1`.

FInally, in `NGBoost Model 3`, I run NGBoost with 1000 estimators, but turn on early stopping. It's not surprise that this model performs differently than `NGBoost Model 1`, but I would expect the result on the test data to be better, not worse (because I have passed the test data in as the validation data here, so there is leakage). Instead, it resembles `NGBoost Model 2` in terms of performance.

As a comparison at the end, I've also presented three examples with LightGBM to show the behavior I would expect to see with stock, validation, and early stopping models respectively. As you see, `LightGBM Model 1' and `LightGBM Model 2` have the same performance, while `LightGBM Model 3` performs slightly better due to early stopping. 

This behavior with the validation data in ngboost makes me conclude I'm not understanding something about the behavior of LightGBM when validation data is passed, so my n_estimators tuning method above may not be workable. Can you advise why I'm seeing different results with NGBoost with validation data (but no early stopping) than I see in LightGBM?

(Tagging @anpatton since I've been collaborating with him on this issue).

**Comment by alejandroschuler** [comment]
@kmedved I think what you're seeing is the combination of two things. First, here is the code that gets called when you predict:

```
    def pred_dist(self, X, max_iter=None):
        """
        Predict the conditional distribution of Y at the points X=x

        Parameters:
            X        : numpy array of predictors (n x p)
            max_iter : get the prediction at the specified number of boosting iterations

        Output:
            A NGBoost distribution object
        """
        if (
            max_iter is not None
        ):  # get prediction at a particular iteration if asked for
            dist = self.staged_pred_dist(X, max_iter=max_iter)[-1]
        elif (
            self.best_val_loss_itr is not None
        ):  # this will exist if there's a validation set
            dist = self.staged_pred_dist(X, max_iter=self.best_val_loss_itr)[-1]
        else:
            params = np.asarray(self.pred_param(X, max_iter))
            dist = self.Dist(params.T)
        return dist
```

In other words, if there a validation set passed to the fit function, then it will record the iteration that minimized the score and use that number of estimators to return the prediction in the future. I remember implementing this behavior, but to be honest I can't remember why I did it like that at the moment.... if you dig into the blame for that code in `ngboost.py` you should be able to find the corresponding commit with what is hopefully a useful commit message.  

That explains why the model fit with early stopping is giving you a similar result: because when passing a validation set, you get the best iteration across all iterations, which is usually very close to the number of estimators identified by early stopping. In fact, your result show what I'd expect from this: a slightly better result when using the absolute best iteration instead of using early stopping. 

This doesn't explain, however, why the RMSE using what is supposed to be the best iteration is *worse* than the result you get when using the full number of estimators fit. What I suspect, however, is that you fill find the results make sense if you instead look at the validation set *score* (which is what ngboost is using to select the best model) instead of RMSE. Although the two should track to some extent, the model with the lowest RMSE from predicted outcomes is not necessarily the model with the lowest negative-log likelihood using the full predicted distribution. 

These two things together, I think, explain your result. I'm a bit tied down at the moment and I can't work on or test the code, but if you can investigate these hypotheses and let me know if they pan out that would be great. I can also change the default prediction behavior of ngboost when validation data are passed in to match that of lightgbm, etc. if we think that makes more sense and doesn't break anything.

**Comment by kmedved** [comment]
Thanks @alejandroschuler - that makes sense on the validation data issue. I suspected it was something like that. I don't have strong thoughts as to whether the LightGBM behavior is better or not. It may be worth adding to the documentation that it's different is all however, since I suspect many users have experience with LightGBM (or xgboost, etc...) and may be surprised otherwise.

I'll need to do some further thinking on how to test your suspicion as to why the RMSE is worse than the full estimators fit. I am surprised by the size of the difference - I understand that they won't track 1 to 1, but I'm getting RMSEs sometimes 2x higher with the early stopping parameter, which is a surprising divergence. 

If you don't mind a somewhat basic question, when you say that ngboost is looking at the validation set **score** to select the best model, can you elaborate by what you mean there? That would help me additional testing on this issue. In other words, what is the score exactly there?

Thanks for the help.

**Comment by alejandroschuler** [comment]
The score is whatever proper scoring rule you are using to evaluate your probabilistic prediction. A loss function, like RMSE, will compare a point prediction to a data point and value how good the prediction is. A scoring rule does the same thing, but instead of comparing a point prediction to a data point, it compares an estimated probability distribution to a data point. The default scoring rule in ngboost is the negative log-likelihood.

There are options in ngboost to tune early stopping based on a rule that isn't the score that is being used to fit the model. I think it's an argument called `val_loss_monitor` or something like that. But this can also be done manually by fitting the model up to the full number of trees, using the staged prediction methods on the validation data to get the predicted distributions and/or point predictions, applying whatever evaluation metric you want to those, and recording the number of iterations that produces the minimum.

**Comment by tonyduan** [comment]
@kmedved Just to add to what Alejandro's said:

I've edited your notebook (I hope you don't mind) to print, for Ex. 1 and Ex. 2, the negative log-likelihood (NLL) in addition to MSE. As expected with test set NLL is lower with early stopping compared to without. What this suggests is the quality of 𝞂 of predicted distributions is higher (as measured by NLL), though at the expense of worse μ of the predicted distributions.

This observation is perhaps not a huge surprise, since we are trying to learn both (μ, 𝞂) from x instead of just μ. In our experiments we note that homoskedastic gradient boosting (i.e. prediction of just μ) tends to result in slightly better MSE:

![Screen Shot 2020-03-13 at 4 01 02 PM](https://user-images.githubusercontent.com/1163598/76665118-dc369400-6543-11ea-927f-602b34d263fe.png)

If MSE is all you're interested in, I'd suggest using the `NormalFixedVar` distribution instead.

I agree that perhaps the interface for early stopping can be made less confusing. Thanks for the feedback.

**Comment by kmedved** [comment]
Thank you both @alejandroschuler and @tonyduan - this is exactly what I was looking for. It seems like the divergence between MSE and NLL is simply much greater than I appreciated. This behavior makes sense now.

And thanks again for taking the time to respond to these somewhat basic questions and for writing/maintaining such a great library in the first place.

**Comment by alejandroschuler** [comment]
Just a heads up that I changed the code so that, even when a validation set is passed, in, you need to manually tell `predict` to predict at the number of iterations that minimized the score, as is done in the [user guide](https://stanfordmlgroup.github.io/ngboost/2-tuning.html).

**Comment by alejandroschuler** [comment]
in other words, now your model 1 and model 2 should be the same

**Comment by kmedved** [comment]
Great. I think that behavior makes a bit more sense, if for no other reason than to more closely align with other libraries.
